src/librbd/deep_copy/ObjectCopyRequest.cc

```diff
@@ -760,8 +765,9 @@ void ObjectCopyRequest<I>::compute_zero_ops() {
 
         if (object_extent.offset + object_extent.length >= end_size) {
           // zero interval at the object end
-          if ((object_extent.offset == 0 && hide_parent) ||
-              (object_extent.offset < prev_end_size)) {
+          if ((object_extent.offset == 0) ||
```

Comment:
@trociny , you had brought up removing these conditions earlier. I am inclined to agree as I don't think it really provides much benefit. What do you think about just updating m_snapshot_sparse_bufferlist without any of these checks?

Reply:
Just doing it unconditionally still looks like a good idea to me. I think it is worth a try.   I hope I have not missed something why we actually need this condition.

Reply:
The test case that fails:
bin/rbd --cluster site-a create -s 12M data/img1
bin/rbd --cluster site-a mirror image enable data/img1 snapshot
bin/rbd map --cluster site-a -t nbd data/img1
xfs_io -d -c 'pwrite -b 4M 0 12M' /dev/nbd0
bin/rbd --cluster site-a du data/img1
bin/rbd --cluster site-a mirror image snapshot data/img1
blkdiscard -o 2M -l 4M /dev/nbd0
bin/rbd --cluster site-a du data/img1
bin/rbd --cluster site-a mirror image snapshot data/img1


Instead of just zeroing out the first 2MB of object oid=1, the entire object is zeroed out in the secondary image.
This is because of the check here:
https://github.com/ceph/ceph/blob/6adc03ac3718e0fe36fb2fc7c2432211158b407b/src/librbd/deep_copy/ObjectCopyRequest.cc#L763
In this case, end_size is 0 due to the incremental snap processing(it doesn't know that the object is fully written to because it does not have the prev snap data), so the if block is executed. This sets the zeroed extents to object_size - offset, hence zeroing out the entire object.

Without the end_size, we cannot determine whether the op is a truncate. 

I was thinking of getting the data for the full snap list each time and only processing the subset. Does that sound like a valid approach?

Reply:
Yes, doing the full snap list looks like a valid approach to me. I believe I suggested it somewhere in this or the previous PR (or was just thinking to suggest?).

Reply:
It looks to me like a good idea to make `ObjectListSnapsRequest` optionally return "previous" end_size, i.e. add `uint64_t *prev_end_size=nullptr` arg, and fill it if it is not null. @idryomov do you have an opinion?

Reply:
Generally, I'm not a fan of optional arguments like that because they tend to be fragile.  I would like the new argument to have clear semantics such that it makes sense in all cases.  So let's define that first -- are you thinking about something like "the size of the object in the starting version"?

Another, potentially better, approach would be to see if there is a way to communicate this through the existing `snapshot_delta` argument such that a truncated extent is reported explicitly, differently from a zeroed extent.  Then `ObjectCopyRequest` wouldn't need to "reconstruct" the truncation at all.

Reply:
It should be "the size of the object in the version before the starting one". Actually I am fine with any approach.

Reply:
> It should be "the size of the object in the version before the starting one".

Is distinguishing between the object that doesn't exist in that version and the object that does exist but has size 0 needed?

Reply:
If I understand your question right then yes, we need to distinguish this. I mean, if the object exists and has size 0 (or any other actually) the `prev_end_size` should be 0 (or the size of this object if it is not 0), but if the object does not exist, the `prev_end_size` should be set from the older version (where the object exists).

Reply:
> It should be "the size of the object in the version before the starting one".

> but if the object does not exist, the `prev_end_size` should be set from the older version (where the object exists).

I might be misreading, but these two statements seem to conflict each other to me.  Does "the version before the starting one" above mean literally the previous snapshot OR an arbitrary preceding snapshot (i.e. the implementation would need to examine all preceding snapshots to make a determination, even in incremental mode)?

And what would `prev_end_size` be set to for an object which never existed in any preceding snapshot?

Reply:
Yes, we need to iterate over all preceding snapshots. And if the object never existed then `prev_end_size` will be 0. Basically we are talking about `prev_end_size` from `ObjectListSnapsRequest::handle_list_snaps` [1], for the iteration when we reached  `first_snap_id`.

Currently in `ObjectCopyRequest::compute_zero_ops` we just set it 0 initially, which worked for "full deep copy" but does not work properly  when we copy only a subset of snapshots. One solution would be to do full snap list in `ObjectCopyRequest` and calculate `prev_end_size` correctly, but we could just reuse the result from `ObjectListSnapsRequest`, which already does this calculation.

https://github.com/ceph/ceph/blob/main/src/librbd/io/ObjectRequest.cc#L897

Reply:
`ObjectListSnapsRequest` tracks not just size, but also existence -- these are separate.  I'm concerned about vague semantics if only `prev_end_size` is exposed to `ObjectCopyRequest`, because `ObjectCopyRequest` tries to reconstruct existence too (`object_exists`).

Reply:
> And if the object never existed then `prev_end_size` will be 0.

... just as it would if the object showed up with size 0 if one of the early snapshots and hasn't been touched since.  It's this sort of ambiguities that I'm wary about.

Reply:
The ObjectSnapListRequest computations will always start from snapid 0 even if the deep copy sends a non-zero snart snap id. It strips out that information before returning the snapshot_delta.
The available info for a particular snapshot includes the end_size (the size of the object for the current snap), exists (if the object exists) and prev_end_size which is the size of the object in the preceding snapshot diff.
The deep copy code is trying to compute this information and sometimes getting it wrong , even though this information is already available in the ObjectSnapListRequest.

We can get all this information very easily by returning this as part of the list_snaps result.

The current fix treats zeroed extents returned in the snapshot_delta as truncate ops as that is the only case when SPARSE_EXTENT_STATE_ZEROED entries are created.

Reply:
@trociny , am I right in thinking that zeroed extents will never be returned by ObjectListSnapsRequest for the parent ?
https://github.com/ceph/ceph/blob/2c43e1d3aa716a230d0cc1079ed7f04fb96b2e74/src/librbd/io/ObjectRequest.cc#L986

Reply:
> `ObjectListSnapsRequest` tracks not just size, but also existence -- these are separate. I'm concerned about vague semantics if only `prev_end_size` is exposed to `ObjectCopyRequest`, because `ObjectCopyRequest` tries to reconstruct existence too (`object_exists`).

We could just return all 3 (exists, prev_end_size and end_size)  as part of the list_snaps call.

Reply:
> @trociny , am I right in thinking that zeroed extents will never be returned by ObjectListSnapsRequest for the parent ?

Yes.  Do you see any issues with it?

Reply:
No. Just that compute_zero_ops function checks for it.